File: karatsuba.py
def karatsuba(x,y):
    if len(str(x)) == 1 or len(str(y)) == 1:
        return x*y
    else:
        m = max(len(str(x)),len(str(y)))
        n = m // 2

        a = x // 10**(n)
        b = x % 10**(n)
        c = y // 10**(n)
        d = y % 10**(n)

        z0 = karatsuba(b,d)
        z1 = karatsuba((a+b),(c+d))
        z2 = karatsuba(a,c)

        return (z2 * 10**(2*n)) + ((z1 - z2 - z0) * 10**(n)) + (z0)
print(karatsuba(3141592653589793238462643383279502884197169399375105820974944592,2718281828459045235360287471352662497757247093699959574966967627))

File: karatsuba.java
public class karatsuba {
  public static void main(String args[]) {
    System.out.println(karatsuba(3141592653589793238462643383279502884197169399375105820974944592,2718281828459045235360287471352662497757247093699959574966967627));
  }

  public static int karatsuba(int X, int Y) {
    String x = Integer.toString(X);
    String y = Integer.toString(Y);
    int result = 0;


    for (int i = 0; i < y.length(); i++) {
      int carry = 0;
      String temp = "";


      for (int j = x.length() - 1; j >= 0; j--) {
        int num = Character.getNumericValue(y.charAt(i)) * Character.getNumericValue(x.charAt(j)) + carry;
        if (num > 9 && j > 0) {
          temp = Integer.toString(num % 10) + temp;
          carry = num / 10;
        }
        else {
          temp = Integer.toString(num) + temp;
          carry = 0;
        }
      }
      result *= 10;

      result += Integer.parseInt(temp);
    }
    return result;
  }
}
